### PR TITLE
feat(theme): support multi-level sidebar

### LINKED ISSUE #926: Support multi-level sidebar
### Is your feature request related to a problem? Please describe.
yes

### Describe the solution you'd like

```json
 {
   "text":"level1",
   "collapsible":true,
   "collapsed":false,
   "items":[
      {
         "text":"level1-1",
         "collapsible":true,
         "collapsed":false,
         "items":[
            {
               "text":"level1-1-1 ",
               "link":"/01.1"
            },
            {
               "text":"level1-1-2 ",
               "link":"/01.2"
            }
         ]
      },
      {
         "text":"level1-2",
         "collapsible":true,
         "collapsed":false,
         "items":[
            {
               "text":"level1-2-1 ",
               "link":"/02.1"
            },
            {
               "text":"level1-2-2 ",
               "link":"/02.2"
            }
         ]
      }
   ]
}

```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://vuejs.org/about/coc.html)
- [X] Read the [docs](https://vitepress.vuejs.org).
- [X] Read the [Contributing Guidelines](https://github.com/vuejs/vitepress/blob/main/.github/contributing.md).
- [X] Check that there isn't already an issue that asks for the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by yizhi996**
#737, fixes #926

Preview:

Screen Shot 2022-06-24 at 21 29 49

### CODE REVIEW

**Comment by brc-dd** [comment]
Hi! Did you accidently closed this PR?

**Comment by yizhi996** [comment]
> Hi! Did you accidently close this PR?

I found a css wrong and temporarily closed this PR.

**Comment by brc-dd** [comment]
Ah okay! Kia will review this once he gets some time. By the way, you can also [convert a PR to draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft) while you're working on it.

**Comment by yizhi996** [comment]
> Ah okay! Kia will review this once he gets some time. By the way, you can also [convert a PR to draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft) while you're working on it.

Thanks, I learned.

**Comment by YangYongAn** [comment]
期待该 pr 早日合并

**Comment by SebasG22** [comment]
Looking forward to having this merged!

**Comment by cy920820** [comment]
Early release pls ~~

File: types/default-theme.d.ts [hunk 1]
```diff
@@ -142,6 +142,7 @@ export namespace DefaultTheme {
   export interface SidebarItem {
     text: string
     link: string
+    items?: SidebarItem[]
```
Comment (brc-dd) [inline types/default-theme.d.ts:145]: The UI looks cool. But I guess this link should be made optional if items is provided.

File: src/client/theme-default/components/VPSidebarLink.vue [hunk 2]
```diff
@@ -5,22 +5,40 @@ import { useData } from 'vitepress'
...
-}>()
+  depth?: number
+}>(), {
+  depth: 1
+})
 
-const { page } = useData()
+const { page, frontmatter } = useData()
+
+const maxDepth: number = frontmatter.value.sidebarDepth || Infinity
```
Comment (brc-dd) [inline src/client/theme-default/components/VPSidebarLink.vue:17]: make this one a computed ref

File: types/default-theme.d.ts (hunk 1 above)
Comment (brc-dd) [reply types/default-theme.d.ts:145]: This type would work:

```ts
  export type SidebarItem =
    | { text: string; link: string }
    | { text: string; link?: string; items: SidebarItem[] }
```

But not providing link is messing up some UI.

**Comment by brc-dd** [comment]
@yizhi996 Hey I did some updates. Can you check if current code works fine?

**Comment by titouanmathis** [comment]
I think the `getFlatSideBarLinks` function in `theme-defaults/supports/sidebar.ts` should be recursive, too. It currently only goes one level deep, which seems to break the prev and next footer links resolution when I last tested it.

I think that the following function could work recursively ([demo](https://jsfiddle.net/6ob9t8vL/)):

```js
function getFlatSideBarLinks(sidebar) {
  return sidebar
    .map((item) =>
      Array.isArray(item.items)
        ? [item, ...getFlatSideBarLinks(item.items)]
        : [item]
    )
    .flat();
}
```

I tested it on this branch, and it seems to work (although TypeScript is not happy):

<table>
  <thead>
    <tr>
      <td>Before</td>
      <td>After</td></tr>
  </thead>
  <tbody>
   <tr>
    <td>
      Capture d’écran 2022-07-22 à 21 36 30
    </td>
    <td>
      Capture d’écran 2022-07-22 à 21 37 44
    </td>
   </tr>
  </tbody>
</table>

**Comment by brc-dd** [comment]
Ah, yes thanks @titouanmathis. This actually slipped my mind. Will fix this tomorrow. I'm marking the PR as draft for now.

**Comment by yizhi996** [comment]
@brc-dd Thanks for the updates to the PR! now it works with `link` be empty

**Comment by yizhi996** [comment]
@brc-dd @titouanmathis I have fixed prev and next footer links problem, please check it works fine, thanks.

**Pull request merged**